### PR TITLE
TypeScript & React/Flux TodoMVC Example

### PR DESCRIPTION
**Pull request by bparadie**
This PR continues the discussion started in https://github.com/tastejs/todomvc/pull/1196 .

Please consider adding this example to `todomvc`.

This application is a port of the [flux-todomvc](https://github.com/facebook/flux/tree/master/examples/flux-todomvc) example, which is part of Facebook's [Flux](https://github.com/facebook/flux) project.

The original example uses [JSX](http://facebook.github.io/react/docs/jsx-in-depth.html), which is a JavaScript syntax extension that looks similar to XML. 
Transforming JSX to Javascript is pretty easy as James Brantly explained in his recent React.js Conf 2015 talk [Static typing with Flow and TypeScript](http://conf.reactjs.com/schedule.html#static-typing-with-flow-and-typescript). James Brantly's [reactconf](https://github.com/jbrantly/reactconf) uses [webpack](http://webpack.github.io/) and [ts-jsx-loader](https://github.com/jbrantly/ts-jsx-loader). This project on the other hand uses [gulp](http://gulpjs.com/),  [react-tools](https://www.npmjs.com/package/react-tools) and [gulp-browserify](https://www.npmjs.com/package/gulp-browserify) instead.

During the build phase all `*.ts` are first being copied to a temporary `./build` folder and then compiled to Javascript using the Typescript compiler. A second pass then intelligently replaces `React.jsx(...)` with transformed JSX code in every `*.jsx.ts` file. For more details see `./gulpfile.js`.
The build mechanics are all implemented in `./gulpfile.js`. The output folder is `./web`.

This project also supports minification using Google's Closure Compiler.

Here is a live demo:
http://bparadie.github.io/fluxts-todomvc

#### Known issues

- `jshint` on `web/js/bundle.js` and `web/js/bundle.min.js` currently fail, because those files are generated from Typescript source files. For that reason I've added a `tshint` pass.
- running the tests against `typescript-reactflux` currently reports 5 failed tests (25 tests pass). Routing is currently not supported and documented in `knownIssues.js`.

Running `mocha allTests.js --no-timeouts --reporter spec --framework=typescript-reactflux` currently produces this output:

```
  TodoMVC - typescript-reactflux
    When page is initially opened
      ✓ should focus on the todo input field
    No Todos
      ✓ should hide #main and #footer (49ms)
    New Todo
      ✓ should allow me to add todo items (419ms)
      ✓ should clear text input field when an item is added (221ms)
      ✓ should append new items to the bottom of the list (506ms)
      ✓ should trim text input (228ms)
      ✓ should show #main and #footer when items added (223ms)
    Mark all as completed
      ✓ should allow me to mark all items as completed (509ms)
      ✓ should allow me to clear the completion state of all items (551ms)
      ✓ complete all checkbox should update state when items are completed / cleared (561ms)
    Item
      ✓ should allow me to mark items as complete (418ms)
      ✓ should allow me to un-mark items as complete (443ms)
      ✓ should allow me to edit an item (728ms)
      ✓ should show the remove button on hover
    Editing
      ✓ should hide other controls when editing (452ms)
      ✓ should save edits on enter (727ms)
      ✓ should save edits on blur (799ms)
      ✓ should trim entered text (735ms)
      ✓ should remove the item if an empty text string was entered (677ms)
      ✓ should cancel edits on escape (691ms)
    Counter
      ✓ should display the current number of todo items (340ms)
    Clear completed button
      ✓ should display the correct text (443ms)
      ✓ should remove completed items when clicked (531ms)
      ✓ should be hidden when there are no items that are completed (500ms)
    Persistence
      ✓ should persist its data (721ms)
    Routing
      1) should allow me to display active items
      2) should respect the back button
      3) should allow me to display completed items
      4) should allow me to display all items
      5) should highlight the currently applied filter


  25 passing (1m)
  5 failing

```

### CODE REVIEW

**Comment by bparadie** [comment]
The build loop shows this error:
```
> jscs "-c" "../.jscsrc" "../examples/typescript-reactflux" "../examples/typescript-reactflux" "../examples/typescript-reactflux" "../examples/typescript-reactflux" "../examples/typescript-reactflux" "../examples/typescript-reactflux" "../examples/typescript-reactflux" "../examples/typescript-reactflux" "../examples/typescript-reactflux" "../examples/typescript-reactflux"
No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
The build has been terminated
```

This looks like a problem in the build scripts. Why are the last 10 parameters all pointing to  `../examples/typescript-reactflux`?

**Comment by arthurvr** [comment]
> running the tests against typescript-reactflux currently reports 5 failed tests (25 tests pass). Routing is currently not supported and documented in knownIssues.js.


Why exactly? It's kind of a requirement.

**Comment by arthurvr** [comment]
> This project also supports minification using Google's Closure Compiler.

cc @tastejs/todomvc but generally we're against build steps already being integrated in the apps. The example should show the framework and not be cluttered with build steps, let's keep it as easy as possible for people.

**Comment by arthurvr** [comment]
> This looks like a problem in the build scripts. Why are the last 10 parameters all pointing to ../examples/typescript-reactflux?

Hmm indeed, sounds like a problem. @samccone known issue?

**Comment by arthurvr** [comment]
Thanks for working on this!

File: examples/typescript-reactflux/LICENSE [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
```
Comment (arthurvr) [inline examples/typescript-reactflux/LICENSE]: What's the deal with the app-specific license? Generally we don't really do that.

File: examples/typescript-reactflux/.gitignore [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+!/typings/tsd.d.ts
+!/typings/todomvc/todomvc.d.ts
+!/typings/react-jsx/react-jsx.d.ts
+!/typings/react-tools/react-tools.d.ts
+
+# Node logs
+npm-debug.log
+
+# OSX
+.DS_Store
```
Comment (arthurvr) [inline examples/typescript-reactflux/.gitignore]: things like `npm-debug.log` and `.DS_Store` should be in your global gitignore.

File: examples/typescript-reactflux/.gitignore [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Generated folders 
+build/
+node_modules/
```
Comment (arthurvr) [inline examples/typescript-reactflux/.gitignore]: don't you use node_modules for front-end dependencies, then? That's how all other example have been doing it, is there something specific preventing it?

File: examples/typescript-reactflux/README.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+## Running
+
+You must have [node.js](https://nodejs.org/) installed on your computer, which includes [npm](https://www.npmjs.org/). From the root project directory run these commands from the command line:
+
+```
+npm install
+```
+
+That will install all dependencies and build `public/js/bundle.js`, which is used by `public/index.html`.
+Once you've built the project you can open `public/index.html` in your browser, or simply run `npm start`.
```
Comment (arthurvr) [inline examples/typescript-reactflux/README.md]: Does this mean that running this is not possible on the live `todomvc.com` site? That's kind of a requirement. I also think the examples should run directly out of the box (without installation or build step). That's how all other examples have been doing it as well.

File: examples/typescript-reactflux/README.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+> Flux is the application architecture that Facebook uses for building client-side web applications. It complements React's composable view components by utilizing a unidirectional data flow.  
+
+> _[Flux - facebook.github.io/flux](https://facebook.github.io/flux/)_
+
+<br>
+
+> React is a Javascript library for building user interfaces. React uses a virtual DOM diff implementation for ultra-high performance.
+
+> _[React - facebook.github.io/react](https://facebook.github.io/react/)_
```
Comment (arthurvr) [inline examples/typescript-reactflux/README.md]: `>` is used to indicate a quote. This is the source of a quote, not a quote :)

File: examples/typescript-reactflux/README.md (hunk 4 above)
Comment (arthurvr) [inline examples/typescript-reactflux/README.md]: well, a build script will of course be required here but then everything should be committed.

File: examples/typescript-reactflux/gulpfile.js [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+  gutil.log('');
+  gutil.log('TASKS:');
+  gutil.log('debug   - builds web/js/bundle.js with runtime JSX renderer.');
+  gutil.log('release - builds web/js/bundle.min.js with expanded JSX.');
+  gutil.log('');
+});
+
+gulp.task('build', ['release']);
+
+gulp.task('default', ['release']);
```
Comment (arthurvr) [inline examples/typescript-reactflux/gulpfile.js]: Many indentation here isn't correct :)

Comment (arthurvr) [inline examples/typescript-reactflux/gulpfile.js]: Also please run JSCS on all javascript files, you'll see a bunch of errors on this one :)

File: examples/typescript-reactflux/gulpfile.js [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+	  debug: true
+	}))
+  .pipe(replace(/function ieKeyFix\(key\)/gm, 'var ieKeyFix = function(key)'))
+  .pipe(replace(/this.module !== module/gm, 'window.module !== module'))
+	.pipe(concat('bundle.js'))
+	.pipe(log())
+	.pipe(gulp.dest( config.web + '/js/'));
+});
+
+gulp.task('minify', function() {
```
Comment (arthurvr) [inline examples/typescript-reactflux/gulpfile.js]: though I understand how a build step might be necessary in this example, minification seems out of scope.

File: examples/typescript-reactflux/gulpfile.js [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+var filter = require('gulp-filter');
+var log = require('gulp-print');
+var replace = require('gulp-replace');
+var reactTools = require('react-tools');
+var browserify = require('gulp-browserify');
+var concat = require('gulp-concat');
+var tsd = require('gulp-tsd');
+var closureCompiler = require('gulp-closure-compiler');
+var runSequence = require('run-sequence');
+var tslint = require('gulp-tslint');
```
Comment (arthurvr) [inline examples/typescript-reactflux/gulpfile.js]: I still don't agree that we should do app-specific linting yet. Let's first clarify on https://github.com/tastejs/todomvc/issues/347.

File: examples/typescript-reactflux/src/actions/TodoActions.ts [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * TodoActions
+ *
+ * Typescript port by Bernd Paradies, May 2015
+ * @see https://github.com/facebook/flux/blob/master/examples/flux-todomvc/js/actions/TodoActions.js
+ */
```
Comment (arthurvr) [inline examples/typescript-reactflux/src/actions/TodoActions.ts]: Did we asked the Facebook people if they're okay with us copying their app? I'd love to hear from them.

File: examples/typescript-reactflux/package.json [hunk 10]
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "fluxts-todomvc",
+  "version": "1.0.0",
+  "description": "Flux TodoMVC Example using Typescript",
+  "author": "Bernd Paradies",
+  "license": "MIT",
+  "repository": "https://github.com/bparadie/fluxts-todomvc",
```
Comment (arthurvr) [inline examples/typescript-reactflux/package.json]: We only really need `private`, scripts and dependencies. That's how we've been doing it on all other apps as well.

File: examples/typescript-reactflux/web/todomvc-app-css/index.css [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+
+@media (max-width: 430px) {
+	#footer {
+		height: 50px;
+	}
+
+	#filters {
+		bottom: 10px;
+	}
+}
```
Comment (arthurvr) [inline examples/typescript-reactflux/web/todomvc-app-css/index.css]: Please include this (and other front-end deps) using a package mananger (preferably npm) similar to the other examples. I don't want hundred different systems to include the same dependencies.

File: examples/typescript-reactflux/web/todomvc-common/base.js [hunk 12]
```diff
@@ -0,0 +1,244 @@
+/* global _ */
+(function () {
+	'use strict';
+
+	/* jshint ignore:start */
```
Comment (arthurvr) [inline examples/typescript-reactflux/web/todomvc-common/base.js]: In fact, why do we need this?

File: learn.json [hunk 13]
```diff
@@ -2130,6 +2133,9 @@
 		}, {
 			"name": "TypeScript & Backbone.js",
 			"url": "examples/typescript-backbone"
+		}, {
+			"name": "TypeScript & React & Flux",
+			"url": "examples/typescript-reactflux/web"
```
Comment (arthurvr) [inline learn.json]: you added it twice?

File: tests/npm-debug.log [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+6 error     <http://github.com/npm/npm/issues>
+7 error System Darwin 14.1.0
+8 error command "/usr/local/bin/node" "/usr/local/bin/npm" "test" "--" "--framework=typescript-reactflux" "--framework=typescript-reactflux" "--framework=typescript-reactflux" "--framework=typescript-reactflux" "--framework=typescript-reactflux" "--framework=typescript-reactflux" "--framework=typescript-reactflux" "--framework=typescript-reactflux" "--framework=typescript-reactflux" "--framework=typescript-reactflux"
+9 error cwd /Users/bparadie/git/research/react/bparadie-todomvc/tests
+10 error node -v v0.10.33
+11 error npm -v 1.4.28
+12 error path /Users/bparadie/git/research/react/bparadie-todomvc/tests/node_modules/--framework=typescript-reactflux/package.json
+13 error code ENOENT
+14 error errno 34
+15 verbose exit [ 34, true ]
```
Comment (arthurvr) [inline tests/npm-debug.log]: please don't include this file

File: examples/typescript-reactflux/.gitignore (hunk 2 above)
Comment (bparadie) [inline examples/typescript-reactflux/.gitignore]: Typo? `should`, or `should not`?

**Comment by bparadie** [comment]
>> running the tests against typescript-reactflux currently reports 5 failed tests (25 tests pass). Routing is currently not supported and documented in knownIssues.js.

> Why exactly? It's kind of a requirement.

"kind of" is interesting. These are the other examples that don't support Routing:

```
        // the following implementations do not support routing
	'TodoMVC - extjs_deftjs, Routing, should allow me to display active items',
	'TodoMVC - extjs_deftjs, Routing, should allow me to display completed items',
	'TodoMVC - extjs_deftjs, Routing, should allow me to display all items',
	'TodoMVC - extjs_deftjs, Routing, should highlight the currently applied filter',
	'TodoMVC - extjs_deftjs, Routing, should respect the back button',
	'TodoMVC - olives, Routing, should allow me to display active items',
	'TodoMVC - olives, Routing, should allow me to display completed items',
	'TodoMVC - olives, Routing, should allow me to display all items',
	'TodoMVC - olives, Routing, should highlight the currently applied filter',
	'TodoMVC - olives, Routing, should respect the back button',
	'TodoMVC - dijon, Routing, should allow me to display active items',
	'TodoMVC - dijon, Routing, should allow me to display completed items',
	'TodoMVC - dijon, Routing, should allow me to display all items',
	'TodoMVC - dijon, Routing, should highlight the currently applied filter',
	'TodoMVC - dijon, Routing, should respect the back button',
	'TodoMVC - duel, Routing, should allow me to display active items',
	'TodoMVC - duel, Routing, should allow me to display completed items',
	'TodoMVC - duel, Routing, should allow me to display all items',
	'TodoMVC - duel, Routing, should highlight the currently applied filter',
	'TodoMVC - duel, Routing, should respect the back button',
	'TodoMVC - knockoutjs_require, Routing, should allow me to display active items',
	'TodoMVC - knockoutjs_require, Routing, should allow me to display completed items',
	'TodoMVC - knockoutjs_require, Routing, should allow me to display all items',
	'TodoMVC - knockoutjs_require, Routing, should highlight the currently applied filter',
	'TodoMVC - knockoutjs_require, Routing, should respect the back button',
	'TodoMVC - angular-dart, Routing, should allow me to display active items',
	'TodoMVC - angular-dart, Routing, should allow me to display completed items',
	'TodoMVC - angular-dart, Routing, should allow me to display all items',
	'TodoMVC - angular-dart, Routing, should highlight the currently applied filter',
	'TodoMVC - angular-dart, Routing, should respect the back button',
	'TodoMVC - typescript-backbone, Routing, should allow me to display active items',
	'TodoMVC - typescript-backbone, Routing, should allow me to display completed items',
	'TodoMVC - typescript-backbone, Routing, should allow me to display all items',
	'TodoMVC - typescript-backbone, Routing, should highlight the currently applied filter',
	'TodoMVC - typescript-backbone, Routing, should respect the back button',
```

**Comment by arthurvr** [comment]
@bparadie https://github.com/tastejs/todomvc/issues/812 :) It's in the app spec and is a requirement, we're just waiting for it to get implemented in the last examples.

File: examples/typescript-reactflux/.gitignore (hunk 2 above)
Comment (arthurvr) [inline examples/typescript-reactflux/.gitignore]: Those files shouldn't be here, those should in your global gitignore :)

**Comment by bparadie** [comment]
>> This project also supports minification using Google's Closure Compiler.

> cc @tastejs/todomvc but generally we're against build steps already being integrated in the apps. The example should show the framework and not be cluttered with build steps, let's keep it as easy as possible for people.

Isn't the whole purpose of this project educating devs how to do things using a particular technology, or framework. If that's the case then the build steps are extremely important in my opinion. 
Nobody is cloning the repo to *use* the todo list app. They want to learn how sh** is implemented. Am I wrong?

**Comment by bparadie** [comment]
> @bparadie #812 :) It's in the app spec and is a requirement, we're just waiting for it to get implemented in the last examples.

Ok, I understand. I'll look at it next weekend then. All of those things are not included in the original Facebook example.

File: examples/typescript-reactflux/LICENSE (hunk 1 above)
Comment (bparadie) [inline examples/typescript-reactflux/LICENSE]: @arthurvr  This is from Facebook. I don't think it would be right to change, or remove the license.

**Comment by arthurvr** [comment]
> They want to learn how sh** is implemented. Am I wrong?

They want to know about the framework, not about gulp or grunt :) If we need a compilation step that's fine, though preferably using npm `scripts` only and it should work fine out of the box.

File: examples/typescript-reactflux/.gitignore (hunk 3 above)
Comment (bparadie) [inline examples/typescript-reactflux/.gitignore]: @arthurvr  I heavily use `node_modules` but only for the build chain. That's why I gitignore it. Most repos do that. I see other examples checking in files from `node_modules` but I consider that bad style.

Comment (arthurvr) [inline examples/typescript-reactflux/.gitignore]: Well, if you're planning to manage front-end dependencies using npm, which you should if it's possible (#1110), ignoring it isn't going to work :)

File: examples/typescript-reactflux/README.md (hunk 4 above)
Comment (bparadie) [inline examples/typescript-reactflux/README.md]: This example runs from `todomvc.com`, or directly in the browser, because I checked in `bundle.min.js`. Those are instructions for devs, who want to copy the example and modify it for their needs. I need to change the folder name from `public` to `web`, though.

File: examples/typescript-reactflux/gulpfile.js (hunk 6 above)
Comment (bparadie) [inline examples/typescript-reactflux/gulpfile.js]: Huh? Should I not indent function body code? The function begins in L319 and ends in L328.
I ran `jshint` on the gulp file but I will try `jscs` as well.

Comment (arthurvr) [inline examples/typescript-reactflux/gulpfile.js]: It (and many other code in this PR) uses two space indentation, we require tabs :)

File: examples/typescript-reactflux/src/actions/TodoActions.ts (hunk 9 above)
Comment (arthurvr) [inline examples/typescript-reactflux/src/actions/TodoActions.ts]: Or at least let them know we did it. Then they don't have to maintain their copy :)

File: examples/typescript-reactflux/gulpfile.js (hunk 7 above)
Comment (bparadie) [inline examples/typescript-reactflux/gulpfile.js]: But that's the interesting part for serious devs! Everybody can `uglify` and compress a JS file. But the tricky part is to minify with Google's closure compiler and [ADVANCED_OPTIMIZATION](https://developers.google.com/closure/compiler/docs/api-tutorial3). I have projects where ADVANCED_OPTIMIZATION gets me compression rates that yield JS file sizes that are up to half of what other 'normal' minifiers produce. 

Now, ADVANCED_OPTIMIZATION is tricky (see `externs`). That's why I wanted to show how it's done in this example.

File: examples/typescript-reactflux/gulpfile.js (hunk 8 above)
Comment (bparadie) [inline examples/typescript-reactflux/gulpfile.js]: For now it's better than no linting at all. I don't think it makes sense to lint minified, or generated JS.

File: examples/typescript-reactflux/src/actions/TodoActions.ts (hunk 9 above)
Comment (bparadie) [inline examples/typescript-reactflux/src/actions/TodoActions.ts]: That's why the license is there. From what I understand, it says 'okay',

File: examples/typescript-reactflux/web/todomvc-app-css/index.css (hunk 11 above)
Comment (bparadie) [inline examples/typescript-reactflux/web/todomvc-app-css/index.css]: But that's exactly what the other examples do, i.e. `angular-dart`. Their examples wouldn't work with the latest css. They checked in files under `node_module`, which essentially freezes the files in time. There is no difference to using copies.

File: examples/typescript-reactflux/gulpfile.js (hunk 7 above)
Comment (arthurvr) [inline examples/typescript-reactflux/gulpfile.js]: > But that's the interesting part for serious devs! Everybody can uglify and compress a JS file. But the tricky part is to minify with Google's closure compiler and ADVANCED_OPTIMIZATION

That's definitely interesting, though users are looking up a todomvc example to see the capabilities of a framework, not to see how they can build build scripts :)

File: learn.json (hunk 13 above)
Comment (bparadie) [inline learn.json]: Yes, one under Typescript and the other one under React.

File: tests/npm-debug.log (hunk 14 above)
Comment (bparadie) [inline tests/npm-debug.log]: Agreed. Why don't you guys add `npm-debug.log` to gitignore?

File: examples/typescript-reactflux/gulpfile.js (hunk 8 above)
Comment (arthurvr) [inline examples/typescript-reactflux/gulpfile.js]: Why don't you take it easy and use a simple npm script? (this whole gulpfile, not only this piece, in fact)

File: learn.json (hunk 13 above)
Comment (arthurvr) [inline learn.json]: yeah but it should have a 'normal' entry as well (like every other example)

File: tests/npm-debug.log (hunk 14 above)
Comment (arthurvr) [inline tests/npm-debug.log]: > Why don't you guys add npm-debug.log to gitignore?

I told you earlier that that's something for a global gitignore :)

**Comment by bparadie** [comment]
>>They want to learn how sh** is implemented. Am I wrong?

>They want to know about the framework, not about gulp or grunt :) If we need a compilation step that's fine, though preferably using npm scripts only and it should work fine out of the box.

I've reduced the whole build chain into one `gulpfile.js`. I thought, that's acceptable. 
My build chain is just a bit more complicated. You can't squeeze it into an npm script.

**Comment by arthurvr** [comment]
> I've reduced the whole build chain into one gulpfile.js. I thought, that's acceptable. 
My build chain is just a bit more complicated. You can't squeeze it into an npm script.

**Comment by arthurvr** [comment]
> I've reduced the whole build chain into one gulpfile.js. I thought, that's acceptable. 
My build chain is just a bit more complicated. You can't squeeze it into an npm script.

It's mainly complicated because most of it is unnecessary. We only require the things necessary to run the app, no minification, for example.

**Comment by bparadie** [comment]
@arthurvr, cc @tastejs/todomvc 

In order to run the app you only need those two CSS files `bundle.min.js` and `index.html`. Why bother posting the Typescript source files, right? Those are not necessary to run the todo list app if that's your main goal. But what would be the point to only present those files that are necessary to run that app? Besides, the code is all summarized in `bundle.min.js` if you really want to know how it's done.

What I like about `todomvc` as a dev community project is that I can learn how those other frameworks work and how you hack them. The best examples in this project are the ones that I can just copy and adjust to some other use case without headaches. When I was bundling up this PR I had in mind to create an example that I would like to study. For me that would definitely include a proper build chain, minification with ADVANCED_OPTIMIZATION, JSX used in a Typescript environment, React and Flux used such it makes sense etc.

I wonder whether I have a complete different picture of what `todomvc` is about.
Is this about learning and sharing knowledge?
Or is this about running a todo list app?

**Comment by urrgur** [comment]
@bparadie You don't get todomvc. It's about the framework, not about a build system. [that are the rules](https://github.com/tastejs/todomvc/commit/adc4c72a370c5ce7c54aeead365f7559a447935b). 

> Why bother posting the Typescript source files, right?

No because you need those for development working and reading code. you doesn't need a gruntfile.

**Comment by bparadie** [comment]
@urrgur My PR uses `gulp`. Same as `grunt` even though [todomvc's root](https://github.com/tastejs/todomvc/blob/master/gulpfile.js) uses it? Do you suggest I should remove `gulpfile.js`? Why so uptight about all this?

**Comment by arthurvr** [comment]
> Do you suggest I should remove gulpfile.js? 

If that's possible just build a really minimal build system using npm scripts. The only thing it really needs to do is to make sure the code is runnable at `todomvc.com/examples/typescript-reactflux`.

> Why so uptight about all this?

Because todomvc examples should stay as simple as possible and focus on the framework. Not on a build system.
